### PR TITLE
[WFCORE-6935] Improve systemd service units

### PR DESCRIPTION
**Pull request by yersan**
Jira issue: https://issues.redhat.com/browse/WFCORE-6935

### CODE REVIEW

File: core-feature-pack/common/src/main/resources/content/bin/systemd/wildfly-domain.service
Comment (yersan) [reply core-feature-pack/common/src/main/resources/content/bin/systemd/wildfly-domain.service:15]: This is just to make `wildfly-domain.conf` an optional file since the Systemd unit is enough to make the server run. `wildfly-domain.conf` can be optionally used to replace the default values but it is not mandatory. A '-' prefix is not used to validate whether the content of the file is correct, it is just to validate if the file exists.

Notice the documentation link is: https://www.freedesktop.org/software/systemd/man/latest/systemd.exec.html#EnvironmentFile=

"""
The argument passed should be an absolute filename or wildcard expression, optionally prefixed with "-", which indicates that if the file does not exist, it will not be read and no error or warning message is logged.
"""

I have no strong opinions about this, right now it aligns with the RPM installations done by the product, so the question is more whether we want to make `wildfly-domain.conf` mandatory to start the server even if the system unit is enough to be able to start the server and you have not added any variable declaration.

One possible advantage of making it mandatory is that you will be sure from the beginning that the file you pretend to use for configuration is read. Otherwise, a weird effect is to be modifying it and see how your server starts but you don't see (or understand) why your environment configuration is not being applied. That could occur if you incorrectly place it in an incorrect place.

File: core-feature-pack/common/src/main/resources/content/bin/systemd/README
Comment (yersan) [reply core-feature-pack/common/src/main/resources/content/bin/systemd/README:15]: I need to check but not sure if that would be fine taking into account that this file contains some other linux commands, I don't want to make them be replaced by maven applying environment properties across this file

**Review by yersan — COMMENTED** [review]
Thanks for the review @jamezp , added some comments related

File: core-feature-pack/common/src/main/resources/content/bin/systemd/README
Comment (jamezp) [reply core-feature-pack/common/src/main/resources/content/bin/systemd/README:15]: That is a good point. I don't see any others with the pattern `${property.name}`, but it's definitely something to consider since that is the linux way.

I was mostly trying to think of a way where we don't have to keep updating the version, but maybe it doesn't matter.

File: core-feature-pack/common/src/main/resources/content/bin/systemd/wildfly-domain.service
Comment (jamezp) [reply core-feature-pack/common/src/main/resources/content/bin/systemd/wildfly-domain.service:15]: This makes sense to me. I was mostly curious on the intent, but you explained it.

**Review by jamezp — APPROVED** [review]
This looks good to me. We can figure out the version thing later and really maybe it doesn't make sense to keep it up to date. After all that is the full version and this is core.

File: core-feature-pack/common/src/main/resources/content/bin/systemd/README
Comment (rpelisse) [inline core-feature-pack/common/src/main/resources/content/bin/systemd/README:42]: To respect Bash's best practises:
`# systemctl start "wildfly-${MODE}.service"`

File: core-feature-pack/common/src/main/resources/content/bin/systemd/README
Comment (rpelisse) [inline core-feature-pack/common/src/main/resources/content/bin/systemd/README:28]: To avoid hard coding /opt/wildfly everywhere, I would suggest defining a WILDFLY_HOME=/opt/wildfly at the beginning.

File: core-feature-pack/common/src/main/resources/content/bin/systemd/README
Comment (rpelisse) [inline core-feature-pack/common/src/main/resources/content/bin/systemd/README:30]: To follow bash's best practices:
```
    # cp "wildfly-${MODE}.conf" /etc/sysconfig/
    # cp "wildfly-${MODE}.service" $(pkg-config systemd --variable=systemdsystemunitdir)
```

File: core-feature-pack/common/src/main/resources/content/bin/systemd/README
Comment (rpelisse) [inline core-feature-pack/common/src/main/resources/content/bin/systemd/README:68]: # rm -f "$(pkg-config systemd --variable=systemdsystemunitdir)/wildfly-${MODE}.service"

File: core-feature-pack/common/src/main/resources/content/bin/systemd/README
Comment (rpelisse) [inline core-feature-pack/common/src/main/resources/content/bin/systemd/README:70]: Should we not add systemctl status wildfy and systemctl start wildfly with an example of the output?

File: core-feature-pack/common/src/main/resources/content/bin/systemd/generate_systemd_unit.sh
Comment (rpelisse) [inline core-feature-pack/common/src/main/resources/content/bin/systemd/generate_systemd_unit.sh:12]: `if [ -z "${JBOSS_SYSTEMD_MODE}" ]; then`

for consistency on top of best practice

File: core-feature-pack/common/src/main/resources/content/bin/systemd/generate_systemd_unit.sh
Comment (rpelisse) [inline core-feature-pack/common/src/main/resources/content/bin/systemd/generate_systemd_unit.sh:18]: `SYSTEMD_FILE="${SCRIPT_DIR}/wildfly-${JBOSS_SYSTEMD_MODE}.service"`

File: core-feature-pack/common/src/main/resources/content/bin/systemd/generate_systemd_unit.sh
Comment (rpelisse) [inline core-feature-pack/common/src/main/resources/content/bin/systemd/generate_systemd_unit.sh:20]: My preference is to add single quote around literals to make it clear to the reader that it's supposed to be fixed characters strings:
`JBOSS_SYSTEMD_GROUP="${3:-'wildfly'}"`

File: core-feature-pack/common/src/main/resources/content/bin/systemd/generate_systemd_unit.sh
Comment (rpelisse) [inline core-feature-pack/common/src/main/resources/content/bin/systemd/generate_systemd_unit.sh:23]: As we have JBOSS_HOME here, we should use it at the beginning, to not repeat /opt/wildfly all the time.

Also: 
`echo "INFO: Using JBOSS_HOME: ${JBOSS_HOME}"`

File: core-feature-pack/common/src/main/resources/content/bin/systemd/generate_systemd_unit.sh
Comment (rpelisse) [inline core-feature-pack/common/src/main/resources/content/bin/systemd/generate_systemd_unit.sh:27]: Interactive shell is a bad idea. I would suggest using opts so that the user can pass a parameters if he wants to generate the systemd file.
We should also aimed at making this script idempotent. (I'll do another review in this regard, after this one)

File: core-feature-pack/common/src/main/resources/content/bin/systemd/generate_systemd_unit.sh
Comment (rpelisse) [inline core-feature-pack/common/src/main/resources/content/bin/systemd/generate_systemd_unit.sh:36]:
```
sed -e "s|@@@JBOSS_SYSTEMD_SERVER_HOME@@@|${JBOSS_HOME}|g" \
    -e "s|@@@JBOSS_SYSTEMD_USER@@@|${JBOSS_SYSTEMD_USER}|g" \
    -e "s|@@@JBOSS_SYSTEMD_GROUP@@@|${JBOSS_SYSTEMD_GROUP}|g" \
    "${SYSTEMD_FILE}.template" > "${SYSTEMD_FILE}"
```

File: core-feature-pack/common/src/main/resources/content/bin/systemd/generate_systemd_unit.sh
Comment (rpelisse) [inline core-feature-pack/common/src/main/resources/content/bin/systemd/generate_systemd_unit.sh:46]: `echo "INFO: Use ${JBOSS_HOME}/bin/systemd/wildfly-${JBOSS_SYSTEMD_MODE}.conf to configure the server environment."`

File: core-feature-pack/common/src/main/resources/content/bin/systemd/generate_systemd_unit.sh
Comment (rpelisse) [inline core-feature-pack/common/src/main/resources/content/bin/systemd/generate_systemd_unit.sh:55]: Be mindful of where to use quote:
`echo sudo journalctl -u $(basename "${SYSTEMD_FILE}")`
This approach will yield a more comprehensible error message if, somehow, a space has been inserted in the variable.

File: core-feature-pack/common/src/main/resources/content/bin/systemd/launch.sh
Comment (rpelisse) [inline core-feature-pack/common/src/main/resources/content/bin/systemd/launch.sh:1]: I would suggest /bin/bash so that we can use bash specific mechanism. It also better because with /bin/sh we should test the script with all sort of shell interpreter.

File: core-feature-pack/common/src/main/resources/content/bin/systemd/README
Comment (yersan) [reply core-feature-pack/common/src/main/resources/content/bin/systemd/README]: Done

File: core-feature-pack/common/src/main/resources/content/bin/systemd/README
Comment (yersan) [reply core-feature-pack/common/src/main/resources/content/bin/systemd/README]: Done

File: core-feature-pack/common/src/main/resources/content/bin/systemd/README
Comment (yersan) [reply core-feature-pack/common/src/main/resources/content/bin/systemd/README]: Done

File: core-feature-pack/common/src/main/resources/content/bin/systemd/README
Comment (yersan) [reply core-feature-pack/common/src/main/resources/content/bin/systemd/README]: Done

File: core-feature-pack/common/src/main/resources/content/bin/systemd/README
Comment (yersan) [reply core-feature-pack/common/src/main/resources/content/bin/systemd/README:71]: At this point, this is the section that describes how to remove the service.

The start / enable / status commands are described in the previous section. There is no sample of status output, though

File: core-feature-pack/common/src/main/resources/content/bin/systemd/generate_systemd_unit.sh
Comment (yersan) [reply core-feature-pack/common/src/main/resources/content/bin/systemd/generate_systemd_unit.sh]: Done

File: core-feature-pack/common/src/main/resources/content/bin/systemd/generate_systemd_unit.sh
Comment (yersan) [reply core-feature-pack/common/src/main/resources/content/bin/systemd/generate_systemd_unit.sh:20]: That would move the quotes to the Systemd unit file as well, is that still valid?

File: core-feature-pack/common/src/main/resources/content/bin/systemd/generate_systemd_unit.sh
Comment (yersan) [reply core-feature-pack/common/src/main/resources/content/bin/systemd/generate_systemd_unit.sh]: > As we have JBOSS_HOME here, we should use it at the beginning, to not repeat /opt/wildfly all the time.

JBOSS_HOME is initialized at the beginning of the script and there is no "/opt/wildfly" path hardcoded in this script, so I don't follow you here.

> Also:
> echo "INFO: Using JBOSS_HOME: ${JBOSS_HOME}"

Done

File: core-feature-pack/common/src/main/resources/content/bin/systemd/generate_systemd_unit.sh
Comment (yersan) [reply core-feature-pack/common/src/main/resources/content/bin/systemd/generate_systemd_unit.sh:27]: > Interactive shell is a bad idea.

It is not expected this file be invoked by a third-party tool, this is just a utility that is expected to be used by a user to make his life easier when he needs to replace the JBOSS_HOME, user:group in the systemd unit files. Asking interactively for confirmation here allows the user to review what he is going to configure, so I think it is useful here.

Would that convince you to leave it as it is?

File: core-feature-pack/common/src/main/resources/content/bin/systemd/generate_systemd_unit.sh
Comment (yersan) [reply core-feature-pack/common/src/main/resources/content/bin/systemd/generate_systemd_unit.sh]: Done

File: core-feature-pack/common/src/main/resources/content/bin/systemd/generate_systemd_unit.sh
Comment (yersan) [reply core-feature-pack/common/src/main/resources/content/bin/systemd/generate_systemd_unit.sh]: Done

File: core-feature-pack/common/src/main/resources/content/bin/systemd/generate_systemd_unit.sh
Comment (yersan) [reply core-feature-pack/common/src/main/resources/content/bin/systemd/generate_systemd_unit.sh]: Done

File: core-feature-pack/common/src/main/resources/content/bin/systemd/generate_systemd_unit.sh
Comment (yersan) [reply core-feature-pack/common/src/main/resources/content/bin/systemd/generate_systemd_unit.sh]: Done

File: core-feature-pack/common/src/main/resources/content/bin/systemd/launch.sh
Comment (yersan) [reply core-feature-pack/common/src/main/resources/content/bin/systemd/launch.sh:1]: We use /sh to maximize the portability now and in the future, it is the same shell used by the launch scripts. Using bash opens the door to not be compatible in the future.

**Review by yersan — COMMENTED** [review]
Thanks for the review @rpelisse , I fixed most of the suggestion and added some minor comments

File: core-feature-pack/common/src/main/resources/content/bin/systemd/launch.sh
Comment (rpelisse) [reply core-feature-pack/common/src/main/resources/content/bin/systemd/launch.sh:1]: Yeah, but do you ensure that CI test the script with different flavor of shell? Unlikely... By using /sh you are opening the door to people filling issue because it does not work with "not Bash". It's generally consider best practises to use explicit /bin/bash for this very reason. As this is a new script, I would take the chance to reduce the scope of incoming issue. That being said, it's not a big deal.

File: core-feature-pack/common/src/main/resources/content/bin/systemd/generate_systemd_unit.sh
Comment (rpelisse) [reply core-feature-pack/common/src/main/resources/content/bin/systemd/generate_systemd_unit.sh:27]: No :) I really think it's a bad idea. 

Bear in mind that this will be used by Ops (or DevOps), they won't expect an interactive shell and will be quite annoyed with it. Also, they are most likely going to automate around this (I've done so in the past, at customers). They may want to automate the deployment of Wildfly based and this and have script (or a tool like Ansible) automatically generate the file on the target system.... 

And right now, it's impossible to because the script is interactive by default. So if you really want to keep it, at least you should provide a mecanism around it.

IMHO, the best approach, that would match what people do in this kind of script, would be to test if the var is defined and act according to it:

```
if  [ -z ${SYSTEMD_FILE} ]; then
   echo "No systemd file provided, not generating one."
   exit 0
fi
```

(last point if somehow you really, really want to keep this interaction thing, I would move into a function, because it's big chunk of code)

Comment (yersan) [reply core-feature-pack/common/src/main/resources/content/bin/systemd/generate_systemd_unit.sh:27]: @rpelisse  Ok, you bring more experience to this from Ansible automation, so I'll remove the interactive question or add a confirm flag for skipping it

Comment (yersan) [reply core-feature-pack/common/src/main/resources/content/bin/systemd/generate_systemd_unit.sh]: @rpelisse I've added a new argument c,--confirm to avoid the interactive mode.

**Comment by yersan** [comment]
Hello @rpelisse could you review this again?

**Review by rpelisse — CHANGES_REQUESTED** [review]
@yersan I'm super sorry to have not reviewed this PR (again) more quickly... I did find a couple of areas I think we should look into before we merge.

**Review by yersan — COMMENTED** [review]
@rpelisse I'm done with this, if there is anything else critital, we would like to merge this as it is now, please take another look, thank you!

File: core-feature-pack/common/src/main/resources/content/bin/systemd/generate_systemd_unit.sh
Comment (rpelisse) [inline core-feature-pack/common/src/main/resources/content/bin/systemd/generate_systemd_unit.sh:15]: I think this might be broken:

`"$(dirname "$(realpath "${0}")")"`

I count 2 opening " and 3 closing one. Also, with command substitution double quotes can be omitted (by spec the $() does the same job as ""). So this should be enough, as the most important "" are around the variable replacement, to ensure the result is treated as one arguments (so a whitespace in the file path does NOT break the script):

`$(dirname $(realpath "${0}"))`

File: core-feature-pack/common/src/main/resources/content/bin/systemd/generate_systemd_unit.sh
Comment (rpelisse) [reply core-feature-pack/common/src/main/resources/content/bin/systemd/generate_systemd_unit.sh:27]: Somehow, I can't close this as resolved, but please do!

File: core-feature-pack/common/src/main/resources/content/bin/systemd/README
Comment (rpelisse) [reply core-feature-pack/common/src/main/resources/content/bin/systemd/README:71]: Somehow, I can't close this as resolved, but please do!

File: core-feature-pack/common/src/main/resources/content/bin/systemd/generate_systemd_unit.sh
Comment (rpelisse) [inline core-feature-pack/common/src/main/resources/content/bin/systemd/generate_systemd_unit.sh:24]: I did not saw it on the first pass, but all the variables in the args treatement are "$XXX" instead of "${XXX}". For consistency purpose and respecting best practices, it might be better to fix that.

File: core-feature-pack/common/src/main/resources/content/bin/systemd/generate_systemd_unit.sh
Comment (rpelisse) [reply core-feature-pack/common/src/main/resources/content/bin/systemd/generate_systemd_unit.sh:94]: Somehow, I can't close this as resolved, but please do!

File: core-feature-pack/common/src/main/resources/content/bin/systemd/launch.sh
Comment (rpelisse) [reply core-feature-pack/common/src/main/resources/content/bin/systemd/launch.sh:1]: Somehow, I can't close this as resolved, but please do!

File: core-feature-pack/common/src/main/resources/content/bin/systemd/generate_systemd_unit.sh
Comment (rpelisse) [reply core-feature-pack/common/src/main/resources/content/bin/systemd/generate_systemd_unit.sh:108]: I think I did not get my point across, but that's fine, it's not a big deal. We can fine tuned that later. I may simply do a PR with my changes and we can discuss directly. For now, good enough, feel free to close it!

File: core-feature-pack/common/src/main/resources/content/bin/systemd/launch.sh
Comment (rpelisse) [inline core-feature-pack/common/src/main/resources/content/bin/systemd/launch.sh:2]: Another missing "${" that I failed to spot the first time :(

File: core-feature-pack/common/src/main/resources/content/bin/systemd/wildfly-domain.service.template
Comment (rpelisse) [reply core-feature-pack/common/src/main/resources/content/bin/systemd/wildfly-domain.service.template:11]: Yeah, we just need to be careful to not update only ONE of them whenever there is a fix to do in them. You can close!

File: core-feature-pack/common/src/main/resources/content/bin/systemd/wildfly-standalone.service
Comment (rpelisse) [reply core-feature-pack/common/src/main/resources/content/bin/systemd/wildfly-standalone.service:15]: It's more to be aligned with the default we picked for the Ansible collection, but you are right, it's fine! You can close this!

File: core-feature-pack/common/src/main/resources/content/bin/systemd/wildfly-standalone.service.template
Comment (rpelisse) [reply core-feature-pack/common/src/main/resources/content/bin/systemd/wildfly-standalone.service.template:14]: Somehow, I can't close this as resolved, but please do!

**Review by rpelisse — CHANGES_REQUESTED** [review]
I think I spot a potential bug, that's why I'm putting this in "Request Changes". I can run shellcheck right now, but I suspect if you do, it will confirm the issue I found. Apart from that (and a couple missing "${xxx}" around variables), I think we're good!

File: core-feature-pack/common/src/main/resources/content/bin/systemd/generate_systemd_unit.sh
Comment (yersan) [reply core-feature-pack/common/src/main/resources/content/bin/systemd/generate_systemd_unit.sh]: Done

File: core-feature-pack/common/src/main/resources/content/bin/systemd/generate_systemd_unit.sh
Comment (yersan) [reply core-feature-pack/common/src/main/resources/content/bin/systemd/generate_systemd_unit.sh]: Done

File: core-feature-pack/common/src/main/resources/content/bin/systemd/launch.sh
Comment (yersan) [reply core-feature-pack/common/src/main/resources/content/bin/systemd/launch.sh]: Done

**Review by yersan — COMMENTED** [review]
@rpelisse I've followed your latest review, Could you check again and approve if you are fine? thansk!

**Review by rpelisse — APPROVED** [review]
LGTM!

**Comment by yersan** [comment]
Thanks @rpelisse and @jamezp

**Pull request merged**
